File: Mavim-services/Topic/Mavim.Manager.Api.Topic.Repository/v1/Mappers/Factory/FieldMapperFactory.cs
using Mavim.Libraries.Authorization.Interfaces;
using Mavim.Manager.Api.Topic.Repository.Interfaces.v1.enums;
using Mavim.Manager.Api.Topic.Repository.v1.Mappers.Abstract;
using Mavim.Manager.Model;
using Microsoft.Extensions.Logging;
using Microsoft.FeatureManagement;
using System;
using System.ComponentModel;
using System.Threading.Tasks;

namespace Mavim.Manager.Api.Topic.Repository.v1.Mappers.Factory
{
    public class FieldMapperFactory
    {
        private readonly IMavimDbDataAccess _dataAccess;
        private readonly ILoggerFactory _loggerFactory;
        private readonly IFeatureManager _featureManager;
        private readonly string hyperlinkFeature = "Hyperlink";
        private readonly string fieldOpenLocationFeature = "FieldOpenLocation";

        public FieldMapperFactory(IFeatureManager featureManager, ILoggerFactory loggerFactory, IMavimDbDataAccess dataAccess)
        {
            _dataAccess = dataAccess ?? throw new ArgumentNullException(nameof(dataAccess));
            _featureManager = featureManager ?? throw new ArgumentNullException(nameof(featureManager));
            _loggerFactory = loggerFactory ?? throw new ArgumentNullException(nameof(loggerFactory));
        }

        internal async Task<FieldMapperBase> GetFieldMapper(FieldType fieldType)
        {
            if (!Enum.IsDefined(typeof(FieldType), fieldType))
                throw new InvalidEnumArgumentException(nameof(fieldType), (int)fieldType, typeof(FieldType));

            return fieldType switch
            {
                FieldType.Unknown => throw new NotImplementedException(),
                FieldType.Text => (FieldMapperBase)new TextFieldMapper(),
                FieldType.MultiText => new MultiTextFieldMapper(),
                FieldType.Number => new NumberFieldMapper(),
                FieldType.MultiNumber => new MultiNumberFieldMapper(),
                FieldType.Decimal => new DecimalFieldMapper(),
                FieldType.MultiDecimal => new MultiDecimalFieldMapper(),
                FieldType.Boolean => new BooleanFieldMapper(),
                FieldType.Date => new DateFieldMapper(),
                FieldType.MultiDate => new MultiDateFieldMapper(),
                FieldType.List => new ListFieldMapper(),
                FieldType.Relationship => await _featureManager.IsEnabledAsync(fieldOpenLocationFeature) ? new RelationshipFieldMapper(_dataAccess) : new RelationshipFieldMapperOld(_dataAccess),
                FieldType.MultiRelationship => await _featureManager.IsEnabledAsync(fieldOpenLocationFeature) ? new MultiRelationshipFieldMapper(_dataAccess) : new MultiRelationshipFieldMapperOld(_dataAccess),
                FieldType.RelationshipList => new RelationshipListFieldMapper(),
                FieldType.MultiHyperlink => new MultiHyperlinkFieldMapper(_loggerFactory),
                FieldType.Hyperlink => new HyperlinkFieldMapper(_loggerFactory),
                _ => throw new ArgumentOutOfRangeException()
            };
        }

        internal async Task<FieldMapperBase> GetFieldMapper(ISimpleField simpleField)
        {
            if (simpleField == null) throw new ArgumentNullException(nameof(simpleField));

            bool? isMulti = null;
            if (simpleField is IMultiValueSimpleField)
                isMulti = true;
            if (simpleField is ISingleValueSimpleField)
                isMulti = false;
            if (isMulti == null)
                throw new Exception("Expected ISingleValueSimpleField or IMultiValueSimpleField");

            switch (simpleField.FieldDefinition.VarType)
            {
                case FieldVarType.Text:
                    if (await _featureManager.IsEnabledAsync(hyperlinkFeature))
                    {
                        var textFieldDefinition = simpleField.FieldDefinition is IFieldTextDefinition ? simpleField.FieldDefinition as IFieldTextDefinition : null;
                        if (textFieldDefinition != null && textFieldDefinition.UriFormat)
                            return ((bool)isMulti ? (FieldMapperBase)new MultiHyperlinkFieldMapper(_loggerFactory) : new HyperlinkFieldMapper(_loggerFactory));
                    }

                    return ((bool)isMulti ? (FieldMapperBase)new MultiTextFieldMapper() : new TextFieldMapper());
                case FieldVarType.Number:
                    return ((bool)isMulti ? (FieldMapperBase)new MultiNumberFieldMapper() : new NumberFieldMapper());
                case FieldVarType.Decimal:
                    return ((bool)isMulti
                        ? (FieldMapperBase)new MultiDecimalFieldMapper()
                        : new DecimalFieldMapper());
                case FieldVarType.Boolean:
                    return new BooleanFieldMapper();
                case FieldVarType.Date:
                    return ((bool)isMulti ? (FieldMapperBase)new MultiDateFieldMapper() : new DateFieldMapper());
                case FieldVarType.List:
                    return new ListFieldMapper();
                case FieldVarType.Relation:
                    return ((bool)isMulti
                        ? (FieldMapperBase)(await _featureManager.IsEnabledAsync(fieldOpenLocationFeature) ? new MultiRelationshipFieldMapper(_dataAccess) : new MultiRelationshipFieldMapperOld(_dataAccess))
                        : (await _featureManager.IsEnabledAsync(fieldOpenLocationFeature) ? new RelationshipFieldMapper(_dataAccess) : new RelationshipFieldMapperOld(_dataAccess)));
                case FieldVarType.RelationList:
                    return new RelationshipListFieldMapper();
                case FieldVarType.Unknown:
                    //case FieldVarType.Integer:
                    //case FieldVarType.Time:
                    //case FieldVarType.InternalFile:
                    // TODO: add logging: WI:19087
                    throw new NotImplementedException();
                default:
                    // returning null now because we don't want to throw an exception when a fieldtype does not have a mapper
                    return null;
            }
        }
    }
}

File: Mavim-apps/Mavim.frontend/projects/mavim-app/src/shared/field/factories/fieldmapperfactory.ts
import {
	BooleanFieldMapper,
	ListFieldMapper,
	NumberFieldMapper,
	MultiNumberFieldMapper,
	DateFieldMapper
} from '../mappers';
import { FieldMapper } from '../mappers/abstract';
import { Field } from '../models/field.model';
import { ApiField, APIFieldValueType } from '../models/api';
import { mapToApiFieldType } from '../mappers/fieldvaluetype';
import {
	TextFieldMapper,
	MultiTextFieldMapper,
	RelationshipFieldMapper,
	MultiDateFieldMapper,
	MultiRelationshipFieldMapper,
	RelationshipListFieldMapper,
	DecimalFieldMapper,
	MultiDecimalFieldMapper,
	HyperlinkFieldMapper,
	MultiHyperlinkFieldMapper
} from '../mappers';
import { FieldValueType } from '../models/fieldvaluetype.enum';

export class FieldMapperFactory {
	public static mapToField(field: ApiField): Field {
		const mapper = this.getFieldMapper(field.fieldValueType);
		if (!mapper) {
			return undefined;
		}

		return mapper.getMappedField(field);
	}

	public static mapToApiField(field: Field): ApiField {
		const fieldType = mapToApiFieldType(field.fieldValueType);
		const mapper = this.getFieldMapper(fieldType);
		if (!mapper) {
			return undefined;
		}

		return mapper.getMappedApiField(field);
	}

	public static mapToPatchUrlType(fieldValueType: FieldValueType): string {
		const fieldType = mapToApiFieldType(fieldValueType);
		const mapper = this.getFieldMapper(fieldType);
		if (!mapper) {
			return undefined;
		}

		return mapper.getUrlTypeName;
	}

	private static getFieldMapper(fieldType: APIFieldValueType): FieldMapper {
		switch (fieldType) {
			case APIFieldValueType.boolean:
				return new BooleanFieldMapper();
			case APIFieldValueType.date:
				return new DateFieldMapper();
			case APIFieldValueType.multiDate:
				return new MultiDateFieldMapper();
			case APIFieldValueType.relationship:
				return new RelationshipFieldMapper();
			case APIFieldValueType.multiRelationship:
				return new MultiRelationshipFieldMapper();
			case APIFieldValueType.relationshipList:
				return new RelationshipListFieldMapper();
			case APIFieldValueType.list:
				return new ListFieldMapper();
			case APIFieldValueType.text:
				return new TextFieldMapper();
			case APIFieldValueType.multiText:
				return new MultiTextFieldMapper();
			case APIFieldValueType.number:
				return new NumberFieldMapper();
			case APIFieldValueType.multiNumber:
				return new MultiNumberFieldMapper();
			case APIFieldValueType.decimal:
				return new DecimalFieldMapper();
			case APIFieldValueType.multiDecimal:
				return new MultiDecimalFieldMapper();
			case APIFieldValueType.hyperlink:
				return new HyperlinkFieldMapper();
			case APIFieldValueType.multiHyperlink:
				return new MultiHyperlinkFieldMapper();
			default:
				return undefined;
		}
	}
}
